### PR TITLE
save method in model fixes payload, validator catches the rest

### LINKED ISSUE #9078: Rancher UI defined roletemplate rules for nonResourceURLs incorrectly have apiGroups definition present, preventing creation of the role in clusters when granted to users
**What kind of request is this (question/bug/enhancement/feature request):**

Bug

**Steps to reproduce (least amount of steps as possible):**

1. Provision a Rancher v2.4.0+ instance with a single all role node custom cluster (I used https://github.com/superseb/tf-do-rancher2/)
2. Under Global -> Security Roles -> Clusters, click `Add Cluster Role`
3. Add a single Resource Grant with `Get` on the Non-Resource URL `/metrics` and remove the `*` in the API Groups field.
4. Grant this new cluster role to a user on the custom cluster.

**Result:**

The roletemplate generated by the UI has a blank, rather than absent `"apiGroups": [""]`, entry, per the following example:

```json
{

    "administrative": false,
    "annotations": {
        "cleanup.cattle.io/rtUpgradeCluster": "true",
        "lifecycle.cattle.io/create.mgmt-auth-roletemplate-lifecycle": "true"
    },
    "baseType": "roleTemplate",
    "builtin": false,
    "clusterCreatorDefault": false,
    "context": "cluster",
    "created": "2020-12-02T16:07:01Z",
    "createdTS": 1606925221000,
    "creatorId": "user-cpgf6",
    "description": "",
    "external": false,
    "hidden": false,
    "id": "rt-h759b",
    "labels": {
        "cattle.io/creator": "norman"
    },
    "links": {
        "clusterRoleTemplateBindings": "…/v3/clusterRoleTemplateBindings?roleTemplateId=rt-h759b",
        "clusters": "…/v3/clusters?defaultClusterRoleForProjectMembers=rt-h759b",
        "projectRoleTemplateBindings": "…/v3/projectRoleTemplateBindings?roleTemplateId=rt-h759b",
        "remove": "…/v3/roleTemplates/rt-h759b",
        "self": "…/v3/roleTemplates/rt-h759b",
        "update": "…/v3/roleTemplates/rt-h759b"
    },
    "locked": false,
    "name": "kube-apiserver metrics",
    "projectCreatorDefault": false,
    "rules": [
        {
            "apiGroups": [
                ""
            ],
            "nonResourceURLs": [
                "/metrics"
            ],
            "type": "/v3/schemas/policyRule",
            "verbs": [
                "get"
            ]
        }
    ],
    "type": "roleTemplate",
    "uuid": "f12c71c5-2c80-44c9-9878-12a2be88e81c"

}
```

As a result of this incorrectly present `apiGroups` definition, for a `nonResourceUrls` rule the clusterole creation will fail when the custom role is granted to a user, with the following error in the rancher server logs:

```
2020/12/02 16:08:07 [INFO] [mgmt-auth-crtb-controller] Updating clusterRoleBinding clusterrolebinding-bnxwk for cluster membership in cluster c-dpvk7 for subject u-fqrkx
2020/12/02 16:08:07 [INFO] Creating clusterRole for roleTemplate kube-apiserver metrics (rt-h759b).
2020/12/02 16:08:07 [ERROR] ClusterRoleTemplateBindingController c-dpvk7/crtb-25kdp [cluster-crtb-sync] failed with : couldn't ensure roles: couldn't create clusterRole rt-h759b: ClusterRole.rbac.authorization.k8s.io "rt-h759b" is invalid: rules[0].nonResourceURLs: Invalid value: []string{"/metrics"}: rules cannot apply to both regular resources and non-resource URLs
```

Same error occurs if you leave the default API Group `*` definition in the form. Removing the `apiGroups` definition from the rule in the role via the API browser, the role and binding is successfully created in the custom cluster.

**Other details that may be helpful:**

When entering a `Non-Resource URL` in the Add Role UI form, the `Resource` field is automatically disabled. The same should occur the API Groups field and no `apiGroups` array should be submitted to the API by the UI for this rule.

The current workaround is to create the role via the API directly.

**Environment information**
- Reproduced with Rancher v2.4.0, v2.4.10 and v2.5.3

### PR DESCRIPTION
**Pull request by Sean-McQ**
### Summary
Fixes #9078 

### Technical notes summary
Implemented form validation to provide user feedback when they have an invalid rule in their roletemplate. customValidationRules in `management.cattle.io.roletemplate` and `management.cattle.io.globalrole` looked to be broken as they were referencing a non-static method on another model which doesn't work. Changed them to what they had been prior to that breaking change and it seems to work now.

### Areas or cases that should be tested
Add a global role and a cluster role
On both, fill out both Resource and Non-Resource URLs
Notice validation message
Notice "Create" button is disabled

### Areas which could experience regressions
None that I can think of, the only orthogonal code change made here was to code that was in a broken state.

### Screenshot/Video
image
image

### CODE REVIEW

File: shell/models/management.cattle.io.roletemplate.js [hunk 1]
```diff
@@ -185,7 +185,16 @@ export default class RoleTemplate extends SteveDescriptionModel {
...
+    for (const rule of norman.rules) {
+      if (rule.nonResourceURLs.length) {
+        delete rule.resources;
+        delete rule.apiGroups;
+      } else {
+        delete rule.nonResourceURLs;
+      }
+    }
+
+    norman.save();
```
Comment (richard-cox) [inline shell/models/management.cattle.io.roletemplate.js:197]: as it doesn't return the result, any failures are ignored

**Review by richard-cox — CHANGES_REQUESTED** [review]
This issue was discussed in grooming but you may not have been there. We need to
- understand if grants are need either `resource` + `api group` OR `non-resource urls`
  - this looks to be the case, but needs confirming https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#policyrule-v1-rbac-authorization-k8s-io
- if so, change the ux such that the user can only enter one set or the other

At the moment
- the user can enter any or none of the resource apigroup or non-resource url fields
- for me no validation warning was shown at any point
  - i couldn't track which `roleTemplateRules` was meant to be used. from what i can tell only one is referenced in code
  -  shell/utils/validators/formRules/index.ts  roleTemplateRules or shell/utils/validators/role-template.js roleTemplateRules
- on `Create` if there is a non-resource both resource and api group are removed. we don't know what the user wanted, so are guessing

**Comment by Sean-McQ** [comment]
The docs say `Rules can either apply to API resources (such as "pods" or "secrets") or non-resource URL paths (such as "/api"), but not both.` which seems pretty definitive. APIGroups isn't required but should auto-populate when a resource is selected from the dropdown. At least one verb is required.

One note: The API does allow the creation of a roletemplate without any resource grants and I can't find any documentation indicating that a roletemplate with no resource grants is invalid so I'm not going to disable save on that condition.

**Review by richard-cox — APPROVED** [review]
After playing around i spotted some unrelated issues we should tackle later

- after adding a `Resource` it's impossible to remove. the entire row has to be removed
- cluster --> view role --> `Config` tab --> api groups and non-resource urls input fields are editable (no `mode`?)
- we need to add an addition input field & validation for the policy `resourceNames` (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#policyrule-v1-rbac-authorization-k8s-io)
- after adding a new grant the warning for empty verb is shown. this shouldn't be shown, as an empty grant is currently valid

**Pull request merged**